primitives/message-lane/src/target_chain.rs

```diff
@@ -37,20 +56,43 @@ pub trait SourceHeaderChain<Payload, Fee> {
 	///
 	/// Messages vector is required to be sorted by nonce within each lane. Out-of-order
 	/// messages will be rejected.
-	fn verify_messages_proof(proof: Self::MessagesProof) -> Result<Vec<Message<Payload, Fee>>, Self::Error>;
+	fn verify_messages_proof(proof: Self::MessagesProof) -> Result<Vec<Message<Fee>>, Self::Error>;
 }
 
 /// Called when inbound message is received.
-pub trait MessageDispatch<Payload, Fee> {
+pub trait MessageDispatch<Fee> {
+	/// Decoded message payload type. Valid message may contain invalid payload. In this case
+	/// message is delivered, but dispatch fails. Therefore, two separate types of payload
+	/// (opaque `MessagePayload` used in delivery and this `DispatchPayload` used in dispatch).
+	type DispatchPayload: Decode;
+
 	/// Estimate dispatch weight.
 	///
 	/// This function must: (1) be instant and (2) return correct upper bound
 	/// of dispatch weight.
-	fn dispatch_weight(message: &Message<Payload, Fee>) -> Weight;
+	fn dispatch_weight(message: &DispatchMessage<Self::DispatchPayload, Fee>) -> Weight;
 
 	/// Called when inbound message is received.
 	///
 	/// It is up to the implementers of this trait to determine whether the message
 	/// is invalid (i.e. improperly encoded, has too large weight, ...) or not.
-	fn dispatch(message: Message<Payload, Fee>);
+	fn dispatch(message: DispatchMessage<Self::DispatchPayload, Fee>);
+}
+
+impl<DispatchPayload: Decode, Fee> From<Message<Fee>> for DispatchMessage<DispatchPayload, Fee> {
```

Comment:
I'd put these closer to the `struct` definitions, but that's just the personal taste.